## The influence of chromosomal inversion on genetic variation and clinal patterns in genomic data

Martin Kapun¹

¹ *Natural History Museum of Vienna, Vienna, Austria*

### Introduction 

Chromosomal inversions are structural mutations the result in the reorientation of the gene order in the affected genomic region. The reversal of synteny impedes homologous pairing in heterokaryotypic chromosomes in the nucleus and leads to loop structures at the chromosomal region spanned by the inversion. These characteristic inversion loops (Figure 1) can be examined under the microscope in polytene chromosomes, which are giant thousand-fold replicated chromatids within a nucleus that can be found, for example, in the salivary glands many Drosophilid fruit fly larvae. For these resasons, inversions were the among first mutations ever to be directly studied more than 100 years ago. Inversion are considered to be the result of ectopic recombination among repetitive and palindromic sequences as found in tRNAs, ribosomal genes and transposable elements (TEs). Accordingly, the breakpoints of inversion polymorphisms, which can range from less than 1,000 bp to several million basepairs in length, are often enriched for such repetitive sequences (for a review, see here). The prevalence of inversions that either include or exclude the centromere (pericentric vs. paracentric inversions) in genomes can vary dramatically, even among closely related taxa and may be linked to the number of genomic TEs. For example, in contrast to the fruit fly *Drosophila melanogaster* which contains many inversions that are pervasive and common in many worldwide populations, its sister taxa *D. mauritiana* and *D. sechellia* are basically inversion-free.

 The primary evolutionary effect of inversions is a strong suppression of recombination (but not gene conversion) with standard arrangement chromosomes since crossing-over within the inverted region results in unbalanced gametes that are non-viable. While recombination in paracentric inversions results in acentric and dicentric gametes, crossing-over in pericentric inversions can cause large-scale duplications and deletions in the recombination products. As a result, both the ancestral standard (ST) and the derived inverted (INV) karyotype evolve largely independent, resulting in increased divergence over time. Since most inversions likely result from unique mutation events, the newly formed inversion evolves from a single haplotype of the ancestral standard arrangement and diverges over time by accumulating novel mutations. Accordingly, inversions may strongly influence the genentic variation in the corresponding genomic region and even beyond the breakpoints. While large chromosomal inversions are generally considered to impose an evolutionary burden since they (1) cause inviable recombination products in heterozygous state, (2) may lead to pseudogenization in the breakpoint regions and (3) shift genes to other genomic regions, which may perturb their expression patterns, they may also have adaptive effects. For example, inversions may  <MORE HERE>

![Figure 1](Images/In3RP.jpg)

### (1) SNPs in strong linkage disequilibrium with inversions

Once a inversions originates and persists in a population, novel mutations will appear and bulid up in frequency  

### (2) SNPs in strong linkage disequilibrium with inversions

